### PR TITLE
[iOS]improve (RCTJSErrorFromCodeMessageAndNSError)

### PR DESCRIPTION
**Pull request by BaihaoTian**
First of all,please forgive my poor english. LOL

When I use Promises  as callBack,

 I followed the guide to make a (RCTPromiseRejectBlock),in this RCTPromiseRejectBlock,

from the guide I should add a (NSString)code a (NSString)message and an (NSError) to a  RCTPromiseRejectBlock. 

As what did guide said I just make a NSError with 

```
 NSError *error=[NSError errorWithDomain:@"native stroage error" code:402 userInfo:@{@"data":key}];

``` 

As you can see, I want to transfer some data in this callback.But I just found the (userInfo) of NSError didn't composed to a  RCTPromiseRejectBlock.

So I check out the source code of RCTUtil ,I found this 

```

// TODO: Can we just replace RCTMakeError with this function instead?
NSDictionary<NSString *, id> *RCTJSErrorFromCodeMessageAndNSError(NSString *code,
                                                                  NSString *message,
                                                                  NSError *__nullable error)
{
  NSString *errorMessage;
  NSArray<NSString *> *stackTrace = [NSThread callStackSymbols];
  NSMutableDictionary<NSString *, id> *errorInfo =
  [NSMutableDictionary dictionaryWithObject:stackTrace forKey:@"nativeStackIOS"];

  if (error) {
    errorMessage = error.localizedDescription ?: @"Unknown error from a native module";
    errorInfo[@"domain"] = error.domain ?: RCTErrorDomain;
  } else {
    errorMessage = @"Unknown error from a native module";
    errorInfo[@"domain"] = RCTErrorDomain;
  }
  errorInfo[@"code"] = code ?: RCTErrorUnspecified;
  // Allow for explicit overriding of the error message
  errorMessage = message ?: errorMessage;
  return RCTMakeError(errorMessage, nil , errorInfo);
}


```

I have noticed that this function compose some part of a error callback but except of userInfo,which I want to transfer. So I think it is necessary to add this part of NSError, in case of somebody want to transfer data in RCTPromiseRejectBlock.

With my poor code skill,I just make a approve of this function to achieve this demand.If it can give you some inspiration and help,it would be a honor for me.

```

NSString *const RCTErrorWithNoData = @"NULL";


NSDictionary<NSString *, id> *RCTJSErrorFromCodeMessageAndNSError(NSString *code,
                                                                  NSString *message,
                                                                  NSError *__nullable error)
{
  NSString *errorMessage;
  NSArray<NSString *> *stackTrace = [NSThread callStackSymbols];
  NSMutableDictionary<NSString *, id> *errorInfo =
  [NSMutableDictionary dictionaryWithObject:stackTrace forKey:@"nativeStackIOS"];

  if (error) {
    errorMessage = error.localizedDescription ?: @"Unknown error from a native module";
    errorInfo[@"domain"] = error.domain ?: RCTErrorDomain;
  } else {
    errorMessage = @"Unknown error from a native module";
    errorInfo[@"domain"] = RCTErrorDomain;
  }
  errorInfo[@"code"] = code ?: RCTErrorUnspecified;
 // add this to achieve transfer data in RCTPromiseRejectBlock
  errorInfo[@"userInfo"] = error.userInfo?:RCTErrorWithNoData;

  // Allow for explicit overriding of the error message
  errorMessage = message ?: errorMessage;
  return RCTMakeError(errorMessage, nil , errorInfo);
}
```

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
By analyzing the blame information on this pull request, we identified @nicklockwood and @grabbou to be potential reviewers.

**Comment by facebook-github-bot-7** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-9** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by facebook-github-bot-4** [comment]
@BaihaoTian updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9519/files/c1658066558709676e816f0610ef0cd79f13faf1..100a992b91ebf9396b3769c179ffb30e5c30a04a)

File: React/Base/RCTUtils.m [hunk 1]
```diff
@@ -423,6 +424,8 @@ BOOL RCTClassOverridesInstanceMethod(Class cls, SEL selector)
     errorInfo[@"domain"] = RCTErrorDomain;
   }
   errorInfo[@"code"] = code ?: RCTErrorUnspecified;
+  errorInfo[@"userInfo"] = error.userInfo?:RCTErrorWithNoData;
```
Comment (javache) [inline React/Base/RCTUtils.m]: Don't use `@"NULL"`, instead do `RCTNullIfNil(error.userInfo)`

**Comment by facebook-github-bot-2** [comment]
@BaihaoTian updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9519/files/100a992b91ebf9396b3769c179ffb30e5c30a04a..675ae870803e55c6e354f2ae9f364dc19c22e611)

**Comment by facebook-github-bot-4** [comment]
@BaihaoTian updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9519/files/675ae870803e55c6e354f2ae9f364dc19c22e611..d770695553f4e4fa21fa17d157156019f900e96b)

**Comment by facebook-github-bot-8** [comment]
@BaihaoTian updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9519/files/d770695553f4e4fa21fa17d157156019f900e96b..cc8431804458cb5804d75ce28adc23e488958bfb)

File: React/Base/RCTUtils.m (hunk 1 above)
Comment (BaihaoTian) [inline React/Base/RCTUtils.m]: tks ,I've followed your advice to change my code.

**Comment by facebook-github-bot-9** [comment]
@BaihaoTian updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9519/files/cc8431804458cb5804d75ce28adc23e488958bfb..b4a999c743e7939aedadf8af8f9321a0d7038da1)

**Comment by javache** [comment]
One issue here is that if you have a description key in your userInfo and no `message` argument to this function, the message will be repeated both in the error name as in the additional info. I advise you to use `RCTMakeError` instead for generating errors instead.

**Comment by javache** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-6** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3821334) to review internal test results.

**Pull request closed without merging**